api/notd/token_metadata_processor.py

```diff
@@ -0,0 +1,45 @@
+import json
+import urllib.request
+from core.requester import Requester
+from core.web3.eth_client import EthClientInterface
+
+from notd.model import RetrievedTokenMetadata
+
+
+class TokenMetadataProcessor():
+
+    def __init__(self, requester: Requester, ethClient: EthClientInterface):
+        self.requester = requester
+        self.ethClient = ethClient
+        with open('./contracts/IERC721Metadata.json') as contractJsonFile:
+            erc721MetdataContractJson = json.load(contractJsonFile)
+        self.erc721MetdataContractAbi = erc721MetdataContractJson['abi']
+        self.erc721MetdataUriFunctionAbi = [internalAbi for internalAbi in self.erc721MetdataContractAbi if internalAbi['name'] == 'tokenURI'][0]
+        self.erc721MetadataNameFunctionAbi = [internalAbi for internalAbi in self.erc721MetdataContractAbi if internalAbi['name'] == 'name'][0]
+
+    async def retrieve_token_metadata(self,registryAddress: str,tokenId: str):
+        tokenMetadataUriResponse = await self.ethClient.call_function(toAddress=registryAddress, contractAbi=self.erc721MetdataContractAbi, functionAbi=self.erc721MetdataUriFunctionAbi, arguments={'tokenId': int(tokenId)})
+        tokenMetadataUri = tokenMetadataUriResponse[0]
+
+        if tokenMetadataUri.startswith('ipfs://'):
+            tokenMetadataUriResponse[0] = tokenMetadataUriResponse[0].replace('ipfs://','https://ipfs.io/ipfs/')
+
+        with urllib.request.urlopen(tokenMetadataUriResponse[0]) as response:
```

Comment:
```suggestion
        with urllib.request.urlopen(tokenMetadataUri) as response:
```

Reply:
sorry it took me a while to realise whats going on here . you should not use urllib directly, please use a requester